### PR TITLE
Reject Project & Consent Group

### PR DESCRIPTION
**Pull request by vvicario**
## Testing

---

- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Get a thumb from @rushtong
- [ ] **Submitter**: Merge to develop using github's "Squash and Merge" feature
- [ ] **Submitter**: Delete branch after merge

### CODE REVIEW

File: grails-app/controllers/org/broadinstitute/orsp/api/NewConsentGroupController.groovy [hunk 1]
```diff
@@ -88,4 +88,22 @@ class NewConsentGroupController extends AuthenticatedController {
...
+    def delete() {
+        Issue issue = queryService.findByKey(params.consentKey)
+        if(issue != null) {
+            Collection<StorageDocument> documents = queryService.getDocumentsForProject(params.consentKey)
+            documents?.each {
+                storageProviderService.removeStorageDocument(it, getUser()?.displayName)
+            }
+            def links = ConsentCollectionLink.findAllByProjectKey(params.consentKey)
+            if(links != null) persistenceService.deleteCollectionLinks(links)
+            issueService.deleteIssue(params.consentKey)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/NewConsentGroupController.groovy]: Nit: There's a good amount of business logic here. In general, we should keep controllers thin and move business logic to a service class instead. I think all of this logic could be in `IssueService` instead.

File: grails-app/controllers/org/broadinstitute/orsp/api/ProjectController.groovy [hunk 2]
```diff
@@ -61,6 +63,23 @@ class ProjectController extends AuthenticatedController {
...
+    def delete() {
+        Issue issue = queryService.findByKey(params.projectKey)
+        if(issue != null) {
+            Collection<StorageDocument> documents = queryService.getDocumentsForProject(params.projectKey)
+            documents?.each {
+                storageProviderService.removeStorageDocument(it, getUser()?.displayName)
+            }
+            def links = ConsentCollectionLink.findAllByProjectKey(params.projectKey)
+            if(links != null) persistenceService.deleteCollectionLinks(links)
+            issueService.deleteIssue(params.projectKey)
```
Comment (rushtong) [inline grails-app/controllers/org/broadinstitute/orsp/api/ProjectController.groovy]: And ... see my comment above - move to `IssueService`.

File: grails-app/services/org/broadinstitute/orsp/IssueService.groovy [hunk 3]
```diff
@@ -237,6 +237,21 @@ class IssueService {
...
+            def fundingList = queryService.findFundingsByProject(issue.projectKey)
+            fundingList?.each {
+                it.delete()
+            }
+            Collection<IssueExtraProperty> issueExtraProperties = issue.getExtraProperties()
+            issueExtraProperties?.each {
+                it.delete()
+            }
+        }
+        issue.delete(flush: true)
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/IssueService.groovy]: See https://github.com/broadinstitute/orsp/blob/develop/sql/utility/delete_issue.sql for a full list of things to look at when deleting an issue. And ... thank you for adding this! I run that delete script every other week or so.

File: src/main/webapp/consentGroupReview/ConsentGroupReview.js [hunk 4]
```diff
@@ -201,8 +202,8 @@ class ConsentGroupReview extends Component {
   }
 
   approveRevision = (e) => () => {
-    this.setState({ disableApproveButton: true })
-    const data = { projectReviewApproved: true }
+    this.setState({ disableApproveButton: true });
+    const data = { projectReviewApproved: true };
```
Comment (rushtong) [inline src/main/webapp/consentGroupReview/ConsentGroupReview.js]: Nit: Throughout, no need for the semi-colons.

File: src/main/webapp/projectReview/ProjectReview.js [hunk 5]
```diff
@@ -175,7 +181,9 @@ class ProjectReview extends Component {
   }
 
   isAdmin() {
-    return this.props.isAdmin === "true";
+    User.getUserSession(this.props.sessionUserUrl).then(resp => {
+      this.setState({isAdmin: resp.data.isAdmin});
+    });
```
Comment (rushtong) [inline src/main/webapp/projectReview/ProjectReview.js]: We should consider caching this at the page level. Every time `isAdmin` is called, we make a rest call, which is probably unnecessary.

Comment (rushtong) [reply src/main/webapp/projectReview/ProjectReview.js]: Also, consider moving admin to props. A user's admin status doesn't really change so it doesn't need to be in state. Try to use state only for things that change, and compute as much as possible.

File: build.gradle [hunk 6]
```diff
@@ -119,6 +119,7 @@ dependencies {
     compile group: 'org.codehaus.groovy.modules.http-builder', name: 'http-builder', version: '0.7.1'
     compile group: 'io.github.http-builder-ng', name: 'http-builder-ng-okhttp', version: '1.0.3'
     compile group: 'org.apache.httpcomponents', name: 'httpmime', version: '4.5.6'
+    compile 'org.grails.plugins:gorm-logical-delete:2.0.0.M2'
```
Comment (rushtong) [inline build.gradle]: Once this is merged, there will need to be a lot of updates to the various sql-based queries in `QueryService` and possibly elsewhere. The plugin auto-magically adds a where clause on the deleted field to GORM queries. Raw sql does not do that and we don't do a lot of GORM querying (have been migrating away from that over time).

File: src/main/webapp/projectReview/ProjectReview.js (hunk 5 above)
Comment (vvicario) [reply src/main/webapp/projectReview/ProjectReview.js]: Greg I reverted my changes, If you agree we will refactor user role code related in US https://broadinstitute.atlassian.net/browse/BTRX-533. 
I finish with logical delete, I'll let you know as soon as walter finish with testing.

File: build.gradle (hunk 6 above)
Comment (vvicario) [reply build.gradle]: Greg if you want I can do the  sql-based queries  updates in this branch

Comment (rushtong) [reply build.gradle]: Not necessary for this branch. Updating the sql queries will only affect issues going through the old flow, not the new, so it makes sense to me that this be a separate effort.

Comment (vvicario) [reply build.gradle]: ok. Thanks!

**Comment by vvicario** [comment]
@rushtong Please let me know if you have any other comments. Walter finished with testing

File: grails-app/services/org/broadinstitute/orsp/IssueService.groovy [hunk 7]
```diff
@@ -237,6 +237,13 @@ class IssueService {
         newIssue
     }
 
+    void deleteIssue(String projectKey) {
+        Issue issue = queryService.findByKey(projectKey)
+        if(issue != null) {
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/IssueService.groovy]: nit: spacing
```suggestion
        if (issue != null) {
```

File: grails-app/services/org/broadinstitute/orsp/IssueService.groovy [hunk 8]
```diff
@@ -237,6 +237,13 @@ class IssueService {
...
+            issue.delete(flush: true)
```
Comment (rushtong) [inline grails-app/services/org/broadinstitute/orsp/IssueService.groovy]: Question: Have you tested that this deletes associations (attachments, submissions, extra properties) as well? I've never been very confident of the framework-delegated functions like this.

File: src/main/webapp/consentGroupReview/ConsentGroupReview.js [hunk 9]
```diff
@@ -217,8 +218,12 @@ class ConsentGroupReview extends Component {
     );
   }
 
-  rejectConsentGroup = (e) => () => {
-
+  rejectConsentGroup() {
+    spinnerService.showAll();
+    ConsentGroup.rejectConsent(this.props.rejectConsentUrl, this.props.consentKey).then(resp => {
+      window.location.href = this.getRedirectUrl(this.props.projectKey);
+      spinnerService.hideAll();
```
Comment (rushtong) [inline src/main/webapp/consentGroupReview/ConsentGroupReview.js]: I didn't catch this before. Could there be an error case where the response fails and `spinnerService.hideAll();` is never invoked? Should this be in a separate handler block? I don't remember the syntax, but there should be an error case for this promise that you could invoke this in.

File: src/main/webapp/consentGroupReview/ConsentGroupReview.js [hunk 10]
```diff
@@ -327,6 +332,31 @@ class ConsentGroupReview extends Component {
...
+    } else {   
+      let key = projectKey.split("-");
+      let projectType = '';
+      if (key.length === 3) {
+        projectType = key[1].toLowerCase();
+      } else {
+        projectType = key[0].toLowerCase();
+      }
+      return [this.props.serverURL, projectType, "show", projectKey,"?tab=consent-groups"].join("/");
+  }
```
Comment (rushtong) [inline src/main/webapp/consentGroupReview/ConsentGroupReview.js]: Nit: indentation on a few lines in this method is off.

**Comment by vvicario** [comment]
@rushtong I added logical delete for issue associated tables and I addressed your last comments!

File: src/main/webapp/consentGroupReview/ConsentGroupReview.js [hunk 11]
```diff
@@ -223,6 +223,9 @@ class ConsentGroupReview extends Component {
     ConsentGroup.rejectConsent(this.props.rejectConsentUrl, this.props.consentKey).then(resp => {
       window.location.href = this.getRedirectUrl(this.props.projectKey);
       spinnerService.hideAll();
+    }).catch(error => {
+      spinnerService.hideAll();
+      console.error(error);
```
Comment (rushtong) [inline src/main/webapp/consentGroupReview/ConsentGroupReview.js]: Nice!

**Pull request merged**